src/Std/Sat/AIG/LawfulOperator.lean

```diff
@@ -0,0 +1,155 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henrik Böving
+-/
+import Std.Sat.AIG.Basic
+
+/-!
+The lawful operator framework provides free theorems around the typeclass `LawfulOperator`.
+Its definition is based on section 3.3 of the AIGNET paper.
+-/
+
+namespace Std
+namespace Sat
+
+namespace AIG
+
+/--
+`decls1` is a prefix of `decls2`
+-/
+structure IsPrefix (decls1 decls2 : Array (Decl α)) : Prop where
+  of ::
+    /--
+    The prefix may never be longer than the other array.
+    -/
+    size_le : decls1.size ≤ decls2.size
+    /--
+    The prefix and the other array must agree on all elements up until the bound of the prefix
+    -/
+    idx_eq : ∀ idx (h : idx < decls1.size), decls2[idx]'(by omega) = decls1[idx]'h
+
+/--
+If `decls1` is a prefix of `decls2` and we start evaluating `decls2` at an
+index in bounds of `decls1` we can evaluate at `decls1`.
+-/
+theorem denote.go_eq_of_IsPrefix (decls1 decls2 : Array (Decl α)) (start : Nat) {hdag1} {hdag2}
```

Comment:
This should be called `denote.go_eq_of_isPrefix` (the rule is that in theorem names, each component should start with a lowercase letter). Please check that you adhere to this rule everywhere.